src/KilnUniV3SwapUniv2LP.sol

```diff
@@ -0,0 +1,179 @@
+        emit File(what, data);
+    }
+
+    /**
+        @dev Auth'ed function to update path value
+        @param what   Tag of value to update
+        @param data   Value to update
+    */
+    function file(bytes32 what, bytes calldata data) external auth {
+        if (what == "path") path = data;
+        else revert("KilnUniV3SwapUniv2LP/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /**
+        @dev Auth'ed function to update yen, zen, or base contract derived values
+             Warning - setting `yen` or `zen` as a low value highly increases the susceptibility to oracle manipulation attacks
+        @param what   Tag of value to update
+        @param data   Value to update
+    */
+    function file(bytes32 what, uint256 data) public override auth {
+        if (what == "yen") {
+            require(data > 0, "KilnUniV3SwapUniv2LP/zero-yen");
+            yen = data;
+        }  else if (what == "zen")  {
+            require(data > 0, "KilnUniV3SwapUniv2LP/zero-zen");
+            zen = data;
+        } else {
+            super.file(what, data);
+            return;
+        }
+        emit File(what, data);
+    }
+
+    function _swap(uint256 amount) internal override returns (uint256 swapped) {
+        uint256 _halfIn = amount / 2;
+        bytes memory _path = path;
+        uint256 quote = IQuoter(quoter).quote(sell, buy, _halfIn);
+
+        GemLike(sell).approve(uniV3Router, _halfIn);
+        SwapRouterLike.ExactInputParams memory params = SwapRouterLike.ExactInputParams({
+            path:             _path,
+            recipient:        address(this),
+            deadline:         block.timestamp,
+            amountIn:         _halfIn,
+            amountOutMinimum: quote * yen / WAD
+        });
+        uint256 bought = SwapRouterLike(uniV3Router).exactInput(params);
+
+        GemLike(sell).approve(uniV2Router, _halfIn);
+        GemLike(buy).approve(uniV2Router, bought);
+        (uint256 depositA, uint256 depositB, uint256 liquidity) = UniswapV2Router02Like(uniV2Router).addLiquidity({
+            tokenA:         sell,
+            tokenB:         buy,
+            amountADesired: _halfIn,
+            amountBDesired: bought,
+            amountAMin:     0, // price check is done separately vs quote below
+            amountBMin:     0,
+            to:             receiver,
+            deadline:       block.timestamp
+        });
+        swapped = liquidity;
+
+        uint256 _zen = zen;
+        uint256 depositToQuote = (depositB * WAD / depositA) * WAD / (quote * WAD / _halfIn);
+        require(depositToQuote >= _zen && depositToQuote <= (2 * WAD - _zen),
+                "KilnUniV3SwapUniv2LP/deposit-price-out-of-bounds"
+        );
+
+        // If not all buy tokens were used, send the remainder to the receiver
+        if (bought > depositB) {
+            unchecked { GemLike(buy).transfer(receiver, bought - depositB); }
```

Comment:
This is the only part that still has me a little uncomfortable.

If the purpose of using the kiln is to get rid of sell token in a permissionless way, dropping the buy token back on the user in dusty amounts means they'll have to find another way to get rid of them if what they really want is the LP token.

Need to think about whether it may be possible to query the UniV2 router first and adjust down the amount we're buying so that we can only spend the amount necessary to fill the deposit order in V3.

I think it would be better to leave `sell` token in the contract that could be used in the next trade instead of overbuying `buy` token and having to send it to the deployer.